solo-io/gloo
# Gloo PDB

# Description

Add `PodDisruptionBudget` to `gloo` helm values

## Helm changes
- Update PodDisruptionBudget fields in values.go to have more relevant descriptions which reference k8s documentation and to have string type, which better represents the k8s IntOrString type
  - Update existing Helm tests for GatewayProxy PDB accordingly
- Add support for `gloo.podDisruptionBudget.minAvailable` and `gloo.podDisruptionBudget.maxUnavailable` values
  - If `gloo.podDisruptionBudget` is specified, a PDB is created with those settings which selects on the `gloo` deployment
  - If both are specified, rendering fails, as those values are mutually exclusive in PDBs
  - Add helm tests, including for the mutually exclusive failure case

# Context

It is not currently possible to configure a PDB for the Gloo controller deployment via Helm

## Interesting decisions

The template is based on that which we already have for the Gateway Proxy [here](https://github.com/solo-io/gloo/blob/a3e5552d9164bb2ca84f0ca6e8b721e72632cbb9/install/helm/gloo/templates/8-gateway-proxy-pod-disruption-budget.yaml)
The template for the Gloo deployment is much simpler because there is only ever one deployment and we don't need to specialize logic to handle PDBs associated with different dynamically-named deployments

As with that template, there is an edge case whereby a user could specify `gloo.podDisruptionBudget` without specifying one of the two accepted values (ie `minAvailable` and `maxUnavailable`)
In this case a PDB will be created for the Gloo deployment without any settings, which appears to always have a `False` status with `InsufficientPods` as the reason
I wasn't able to determine what ramifications this might have for the normal operation of the Gloo deployment (see below under Testing steps)
If we'd prefer to be more strict about requiring a valid setting to present in order to create the PDB I'll fix it in both locations

Though the `PodDisruptionBudget` object in `values.go` includes `KubeResourceOverride`, this functionality is not relevant for the Gloo PDB
A line is [generated in docs](https://github.com/solo-io/gloo/pull/8600/files#diff-2baf5c822ecb28b315e303c9d28c5c3a6486b975c91e7ba2d19d7eb05966c882R235) describing this value
I do not expect users to be too confused by this, however if reviewers would prefer, I can create a new `PodDisruptionBudget` struct that omits `KubeResourceOverride` 

It may be overly verbose/redundant to refer to values by their full reference (eg `.Values.gloo.podDisruptionBudget.minAvailable`) every time, however I find this more readable and there isn't much benefit to assigning local variables in such a small and straightforward template

Some improvements have been made to the paradigm set by the Gateway Proxy PDB however they have not been propagated to that existing PDB template in order to keep the scope narrow
I am open-minded if reviewers feel that it is worthwhile to propagate these changes despite scope creep

## Testing steps

I used `helm template` commands to confirm that the output looked right, for example:
```
$ helm template install/helm/gloo --set gloo.podDisruptionBudget.maxUnavailable=3 > template.yaml
```

I subsequently did a local `helm install` and confirmed that the generated PDB correctly referenced the Gloo Deployment
```
$ helm install gloo install/helm/gloo --set gloo.podDisruptionBudget.maxUnavailable=3
```

This results in the following PDB:
```
apiVersion: policy/v1
kind: PodDisruptionBudget
metadata:
...
  name: gloo-pdb
  namespace: default
...
spec:
  maxUnavailable: 3
  selector:
    matchLabels:
      gloo: gloo
status:
  conditions:
  - lastTransitionTime: "2023-08-21T18:14:04Z"
    message: ""
    observedGeneration: 1
    reason: SufficientPods
    status: "True"
    type: DisruptionAllowed
  currentHealthy: 1
  desiredHealthy: 0
  disruptionsAllowed: 1
  expectedPods: 1
  observedGeneration: 1
```
Note the `currentHealthy: 1` field in the status, indicating that the deployment's pods are being watched correctly
Even with a PDB, manual deletes of pods are always allowed, and I wasn't able to determine another method for testing a PDB, however it is sufficient to verify correctness of the Helm values to observe that a PDB with the correct settings is correctly associated with the Gloo deployment
 
## Notes for reviewers

Feel free to pull the branch and verify locally with either `helm template` or `helm install` as I have done

I welcome feedback on the design decisions described above

Note that though the Helm tests required updating to explicitly pass values as strings, in practice it still works to pass the values as ints (ie without quotes)

# Checklist:


BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/8445
